Default enarx trust process (#26)
enarx/rfcs: concepts/default-enarx-trust-process.md

# TBD: Default Enarx trust process
- Authors: [Mike Bursell](mike@p2ptrust.org)
- Status: [PROPOSED](/README.md#proposed)
- Since: 2020-03-17
- Status Note: under discussion
- Start Date: 2020-03-17 (date you started working on this idea)
- Tags: trust
- Addresses: (Issue #)[https://github.com/enarx/enarx/issues/]

TODO: this is curently listed as a "concepts" document, but also
includes design information and implementation requirements.  Need
to decide where it best sits.  Neither "concepts" nor "features"
(as described in the README.md) fit very well.

## Summary

Setting up trust domains and being able to transfer certain entities
between them in central to Enarx.  This RFC describes the default
trust process for Enarx and lays out sets of requirements for
implementations.

## Motivation

One of the key aims of the Enarx project is to allow clients to run
workloads on hosts which are "untrusted".  This RFC builds on previous

> RFCs to describe the default model by which Enarx allows this take

Reviewer: ```suggestion
RFCs to describe the default model by which Enarx allows this to take
```

place, using the capabilities of TEEs, the underlying CPU(s) and
firmware of the host and other parties.

It should be noted that the key party for whom trust is considered
important within the Enarx project is the workload owner - in this case,

> the tenant.  It is trust relationships from this entity to others, and

Reviewer: > trust relationships from this entity to others

I'm not sure how to interpret this. "The trust relationships from this entity's point of view"?

Author: I've tried to clarify this.  Please let me know if I've succeeded!

Reviewer: It's certainly *more* clear now, but I'm still not sure. If what matters is trust *from me to you*, does that mean it's important that I trust you or that you trust me? I can extrapolate this from the context since I already know what's supposed to be going on, but it takes me a minute to think about this phrasing. Could be just me, though.

Author: I've taken some suggested text from @axelsimon which I think helps.

maintenance of the tenant's trust domain, which are always the
priority - any other benefits which may accrue to other parties, are
considered side benefits.

On a different note, this RFC deals mainly with the communications
between the various Enarx components.  Although the Orchestrator is
involved and mentioned in this document, it is not the document's focus.

## Trust state transitions in Enarx

See rfc#0000x for an introduction to trust relationships.
See rfc#0000x for an introduction to trust domains and their application
to Enarx.
See rfc#0000x for an introduction to endorsing authorities, trust anchors
and trust pivots.

rfc#0000x described the following Enarx use case states for the default
Enarx trust model (TODO: check correctness in rfc#0000x and normalise):
 - State 0 - initial state
 - State 1 - pre-attestation
 - State 2 - attested
 - State 3 - provisioned

Non-Enarx use cases were also discussed, but these are irrelevant to this
discussion, and therefore ignored in this RFC.

The changes between states were characterised as the transition of

> particular components between to host domains:

Reviewer: ```suggestion
particular components between trust domains:
```

Author: Fixed.

    - Tenant workload image
    - Tenant workload image repository

#### State 1

 - host owner trust domain
    - host
    - host CPU + CPU firmware
    - Enarx host agent
    - **TEE instance**
    - **Enarx runtime instance (within TEE instance)**
    - Other workloads
 - tenant trust domain
    - Orchestrator
    - Enarx client agent
    - Tenant workload image
    - Tenant workload image repository

#### Changes

The changes between state 0 and state 1 are the creation of a TEE instance
and the loading of the Enarx runtime image into that instance.

The initialisation of the process is prompted by **Orchestrator**, which
communicates with the **Enarx client agent**, which contacts the **Enarx
host agent** over a network connection (which does not need to be
encrypted).  A mechanism (such as a cryptographic nonce), allowing the
Enarx client agent to match sessions and requests, is included in the
initial request.  The **Enarx host agent** then requests the creation,
which is performed by the **host CPU + CPU firmware**.

The **Enarx runtime image** is typically stored on host storage (local or
remote - the distinction is not relevant to the process).  The loading

> process varies between architectures (TODO: is this true?), but is

Reviewer: We do have different code for each loading process.

Author: Not quite what I was asking, which was really "do we believe that the actual process for loading will be different, e.g. one will have direct access to local storage, another listens on a socket, etc.?".

Reviewer: Good questions. : ) I'm curious about the answers....

Reviewer 2: @lkatalin Who do you think might have these "answers" you mention? 🤔

Reviewer: We should definitely check with @npmccallum about whether the loading process varies between architectures.

Once the **TEE instance** has been created, the **host CPU + CPU firmware**
load the **Enarx runtime image** into the **TEE instance**.  The
**Enarx runtime image**, now loaded, is executed, and becomes the
**Enarx runtime instance**, though it is not yet attested.

As part of the initialisation process, the **Orchestrator** provides the
**Enarx client agent** with one of the following:
 1. access to the **Tenant workload image repository** and a reference to
 **Tenant workload image** to be provisioned;
 2. direct access to the **Tenant workload image**, either as a binary in
 memory or in storage;
 3. a reference allowing the **Enarx client agent** to recontact the
 **Orchestrator** in the event of successful attestation.
 
Though the encryption of the channel between the Enarx client agent
and the Enarx host agent is not **required**, the following properties are
considered to be **desirable**:

 - authentication of the Enarx client agent to the Enarx host agent
   - this allows the host to control which parties are permitted to create
   Enarx Keeps.  While this has no direct benefit to the tenant, an
   indirect benefit may be that the host can avoid DoS attacks, thereby

>    maintaining availability for workload placement and execution.

Reviewer: This is an interesting point.

   - some DoS attacks or protocol attacks may be feasible, and mitigated
   by integrity protection.
 - encryption of the protocol handshake and further steps
   - some attacks based around protocol versioning or replays may be
   feasible, and can be avoided by encryption.

For the reasons noted above, encrypting the channel will sometimes be considered
the simplest option.

### State 1->2 change

#### State 1 - pre-attestation

 - host owner trust domain
    - host
    - host CPU + CPU firmware
    - Enarx host agent
    - TEE instance
    - Enarx runtime instance (within TEE instance)
    - Other workloads
 - tenant trust domain
    - Orchestrator
    - Enarx client agent
    - Tenant workload image
    - Tenant workload image repository

#### State 2 - attested

 - host owner trust domain
    - host
    - host CPU + CPU firmware
    - Enarx host agent
    - Other workloads
 - tenant trust domain
    - Orchestrator
    - Enarx client agent
    - **Empty Keep**
    - Tenant workload image
    - Tenant workload image repository

#### Changes

The change between state 1 and state 2 is the transition of the TEE
instance and Enarx runtime image within it to an Enarx Keep.  This

> transition is considered complete once the Keep has been attested.

Reviewer: Calling it a Keep only after attestation is a good distinction.

Author: I think so, too: and it starts out empty.  So:
0. nothing
1. TEE instance
2. Empty Keep (== attested(TEE instance + Enarx runtime))
3. Keep (== attested(TEE instance + Enarx runtime) + workload)

Reviewer 2: Agreed, this makes it easier to explain or "sell" what separates a Keep from a TEE (instance). It's not just a neat word, it's a properly set-up instance, which includes attestation done right.

The measurement of the TEE instance and Enarx runtime image within it
is requested by the **Enarx host agent** and performed by the **host
CPU + CPU firmware**.  The measurement is then transmitted from the
**Enarx host agent** to the **Enarx client agent**.

The **Enarx host agent** checks the attestation measurement database
(TODO: defined in a separate RFC) to ensure that the attestation
measurement is correct.  This process includes checks for:

 - host CPU + CPU firmware validity
 - cryptographic correctness
 - TEE validity
 - valid Enarx runtime image.

If any of these checks fail, the state progression is stopped, and the
**Enarx client agent** logs an error and may send an error message to
the **Enarx host agent**.  The **Enarx client agent** also sends an
error message to the Orchestrator.

If the attestation measurement meets the checking criteria, the **Enarx
client agent** retrieves or derives the session key from the previous

> communication (the exact mechanism being CPU architecture-dependent).  At

Reviewer: ```suggestion
client agent** retrieves or derives the session key from the previous
measurement originating in the **host CPU + CPU firmware** (the exact mechanism being CPU architecture-dependent).  At
```
I would prefer it to be extremely clear that the untrusted host and host agent are not involved in generating the session key and cannot alter it.

Author: Like it.  Fixed.  Also fixed an egregious error at line 220 which suggested that the Enarx host agent checks the database, when it's actually the Enarx client agent.

this point, the state transition is considered complete, and the TEE
instance, loaded with the Enarx runtime image, is now an **Empty Keep**,
which includes a running **Enarx runtime instance**. 

### State 2->3 change

> #### State 2 - pre-attestation

Reviewer: ```suggestion
#### State 2 - attested
```

Author: Fixed (and subsequent cut and paste error for State 3).

 - host owner trust domain
    - host
    - host CPU + CPU firmware
    - Enarx host agent
    - Other workloads
 - tenant trust domain
    - Orchestrator
    - Enarx client agent
    - **Empty Keep**
    - Tenant workload image
    - Tenant workload image repository

> #### State 3 - attested

Reviewer: ```suggestion
#### State 3 - provisioned
```

Author: Fixed.

**Enarx client agent** with one of the following:
 1. access to the **Tenant workload image repository** and a reference to
 **Tenant workload image** to be provisioned;
 2. direct access to the **Tenant workload image**, either as a binary in
 memory or in storage;
 3. a reference allowing the **Enarx client agent** to recontact the
 **Orchestrator** in the event of successful attestation.

The **Enarx client agent** now requires access to the **Tenant workload
image**.

 - In the case of 1. above, the **Enarx client agent** now exercises its
 access to the **Tenant workload image repository** and retrieves the
 **Tenant workload image**.
 - In the case of 2. above, the **Enarx client agent** already has access
 to the **Tenant workload image**.
 - In the case of 3. above, the **Enarx client agent** contacts the
 **Orchestrator**, providing attestation measurement check status
 information, and receives the **Tenant workload image** from the
 **Orchestrator**.  Note that explicit approval should be included in the
 communication to allow the **Orchestrator** to check that the message is
 a "success" and not a "failure": this is a measure to reduce sensitive

>  information leakage in the event of implementation errors.

Reviewer: Great.

> 

Reviewer: ```suggestion
```

Author: Fixed.

The **Enarx client agent** gained access to a session key as part of the
state 1->2 transition.  The **Tenant workload image** must be encrypted
under this session key to be transmitted to the **Empty Keep**.

TODO: check this - particularly authentication pieces.
At some point in the process (undefined in this document), the **Enarx
client agent** was provided with sufficient information to contact what is
now the **Empty Keep**.  This may have been via set-up information from
the **Orchestrator**, as part of the communications with the **Enarx host
agent**, or via "call-back" from the **Empty Keep**: in the latter case,
the **Enarx client agent** must be certain (through cryptographic
authentication) that the **Empty Keep** is the expected entity.  This may
be achieved by the injection of authentication material into the **Empty
Keep** at provision-time, but this must then be tied to the attestation
measurement, and unique to each **Empty Keep** instance.  The session key
established in the state 1->2 transition may be sufficient for this

> purpose.

Reviewer: I'm a bit confused by this paragraph. Are there scenarios in which keeping the same communication pathway as before (client agent -> host agent -> now the Keep) but encrypting with the session key would *not* be good enough, or in which the session key would *not* be sufficient to prove to the client agent that the Keep is the expected entity?

Author: I don't know, and I'm not sure what set of options we want to entertain.  This is one of the paragraphs that I'd like the group to spend some time discussing.

Reviewer 2: I agree. I can feel there might be some not-so-obvious cases here and that we should tread carefully, but right now, the path described by @lkatalin seems fine.

The **Enarx client agent** now creates an encrypted communications channel
to the **Empty Keep**, using the session key established in the state 1->2

> transition.  The **Empty Keep** will be listening for a network connection,

Reviewer: Did we end up deciding that the network connection would actually be in the Keep, as opposed to having the host ferry encrypted communications between the Keep and client? I don't remember now, but will dig up the thread on this to refresh my memory.

Author: I believe that this is what the RFC says, but it's still in draft, awaiting review from @npmccallum ...

Reviewer 2: Did you end up finding the thread, @lkatalin? It could help refresh everyone's memory and inform @npmccallum's review.

Reviewer: I believe I was thinking of this one: https://github.com/enarx/rfcs/issues/4

When a trade-off between the security of the tenant and workload needs
to be made against the security of any other party (including host owner),
the security of the tenant MUST always win.

When measures for a party other than the tenant are being considered,
the general principle should be to choose the highest levels of security,
but any choices MUST NOT compromise tenant security. 

The Tenant Workload MUST never be unencrypted on the network (in transit)
within the Enarx-managed process.

The Tenant Workload MUST never be unencrypted in storage once it has been
transmitted by the Enarx client agent.

The Tenant Workload MUST never be unencrypted in RAM once it has been
transmitted by the Enarx client agent.

The Tenant Workload MUST never be available in unencrypted form on the host.

### Other requirements
The channel between the Enarx client agent and the Enarx host agent MAY be
encrypted.

The Enarx client agent SHOULD authenticate to the Enarx host agent.

Integrity protection of the request SHOULD be applied.

The protocol handshake and further steps SHOULD be encrypted.

If integrity protection is applied, all detected integrity failures SHOULD

> be reported to the Enarx client agent via the Enarx host agent where possible.

Reviewer: This confused me until I realized it is still talking about the integrity of the initial request from client to host agent. Is that right? I read it as integrity protection of the workload.

Author: Clarified.

checked in every attestation measurement check.

All cryptographic validity check failures of the host CPU + CPU firmware
MUST result in an attestation failure.

The cessation of the standard state progression before transmission of
the tenant workload MUST result in non-transmission of the workload from
the Enarx client agent to any component on the host.

All attestation measurement failures MUST be logged by the Enarx host
agent.

All attestation measurement failures SHOULD result in an error message
being sent to the Orchestrator.

The Enarx client agent MUST generate and include session tracking
information in the initial communication with the Enarx host agent.

The session tracking information MUST be included in all communications
between the Enarx client agent and the Enarx host agent until the Keep
is fully provisioned.

The Tenant Workload MUST never be unencrypted on the network (in transit)
within the Enarx-managed process.

The Tenant Workload MUST never be unencrypted in storage once it has been
transmitted by the Enarx client agent.

The Tenant Workload MUST never be unencrypted in RAM once it has been
transmitted by the Enarx client agent.

> The Tenant Workload MUST never be available in unencrypted form on the host.

Reviewer: These are already in the general requirements section.

Author: Yes, I left them in just in case we wish to separate out the general requirements section to another document.  Need to tidy up once we decide.

# TBD: Default Enarx trust process
- Authors: [Mike Bursell](mike@p2ptrust.org)
- Status: [PROPOSED](/README.md#proposed)
- Since: 2020-03-17
- Status Note: under discussion

> - Start Date: 2020-03-17 (date you started working on this idea)

Reviewer 3: ~~(date you started working on this idea)~~

Author: > I think some diagrams would help reduce the amount of prose here. The state machine could be represented as a diagram and then each state could contain prose to describe it.

I think you're right, but I've tried to stick with standard RFC plain text.  I can _try_ to do state machines in ASCII art, I suppose: what do you think?  Alternatively, some pictures could be referenced.

> Furthermore, the process flow description could be factored out into something like an mscgen process flow diagram which would further reduce the amount of prose needed. In many cases we won't eliminate prose here, but I think it would allow for more concision.

Again, text is good and unambiguous, but... pictures are more concise.

> I'm also hesitant to approve an RFC with TODOs still embedded in the text.

Agreed.  I think that all(?) of the TODOs refer to other RFCs awaiting approval.  Any others should probably be issues.

Reviewer 3: For the state machines I was thinking of dot graphs. For example: [source](https://github.com/enarx/enarx/blob/master/docs/attestation/amd/sev/certchain.dot), [output](https://github.com/enarx/enarx/blob/master/docs/attestation/amd/sev/certchain.dot.png).

For the process flow diagrams I was thinking of mscgen graphs. For example: [source](https://github.com/enarx/enarx/blob/master/docs/attestation/amd/sev/process.msc), [output](https://github.com/enarx/enarx/blob/master/docs/attestation/amd/sev/process.msc.png)

I think the visual representation of the process flow or the state machine would just pay back dividends in terms of making sense of it all.

- Tags: trust

> - Addresses: (Issue #)[https://github.com/enarx/enarx/issues/]

Reviewer 3: This field is incomplete.

TODO: this is curently listed as a "concepts" document, but also
includes design information and implementation requirements.  Need
to decide where it best sits.  Neither "concepts" nor "features"

> (as described in the README.md) fit very well.

Reviewer 3: I believe some recent commits have removed the category directory to this repo so this TODO can be removed.

> (as described in the README.md) fit very well.

Reviewer 2: Not a problem anymore, as we are removing the distinction.

## Summary

Setting up trust domains and being able to transfer certain entities
between them in central to Enarx.  This RFC describes the default
trust process for Enarx and lays out sets of requirements for
implementations.

## Motivation

One of the key aims of the Enarx project is to allow clients to run
workloads on hosts which are "untrusted".  This RFC builds on previous
RFCs to describe the default model by which Enarx allows this to take
place, using the capabilities of TEEs, the underlying CPU(s) and

> firmware of the host and other parties.

Reviewer 3: Reword:

"allows this to take place through the use of TEE capabilities supported by the host CPU(s)."

Author: I don't agree: the capabilities are a combination of these three entities.

It should be noted that the key party for whom trust is considered
important within the Enarx project is the workload owner - in this case,

> the tenant.  We should remember that all trust relationships are

Reviewer 3: Reword:

"is the workload owner, also referred to as the tenant"

Author: Disagree - the workload owner may _sometimes_ be the tenant (and is, in this default use case), but will not always be.  We have other trust models to explore.

> the tenant.  We should remember that all trust relationships are

Reviewer 2: ```suggestion
It should be noted that within the context of the Enarx project, trust is 
always considered first and foremost from the perspective of the 
workload owner, that is, the tenant. We should remember that all trust relationships are
```

Trying to simplify a bit that sentence, which works but required me several readings to ensure i understood it correctly.
This suggestion admittedly doesn't express exactly the same idea, but may either be good enough or provide an alternative idea as to how to put the idea forth.

Author: Nice, I've taken this (manually merged).

explicitly uni-directional (though this does not preclude a
corresponding trust relationship in the other direction).  In the case
of Enarx, it is trust relationships from the workload entity to other
entities (primarily the host owner in this document), and maintenance

> of the tenant's trust domain, which are always the priority - any

Reviewer 3: I think we should put a full-stop after the word "priority" and let "Any" be the start of a new sentence.

Author: Agreed.  Fixed.

other benefits which may accrue to other parties, are considered
side benefits.

On a different note, this RFC deals mainly with the communications
between the various Enarx components.  Although the Orchestrator is
involved and mentioned in this document, it is not the document's focus.

## Trust state transitions in Enarx

> See rfc#0000x for an introduction to trust relationships.

Reviewer 2: Currently [#00002](https://github.com/enarx/rfcs/pull/3).

> See rfc#0000x for an introduction to trust domains and their application

Reviewer 2: Currently [#00003](https://github.com/enarx/rfcs/pull/3).

to Enarx.

> See rfc#0000x for an introduction to endorsing authorities, trust anchors

Reviewer 2: Currently [#00005](https://github.com/enarx/rfcs/pull/21).

> and trust pivots.

Reviewer 3: Let's remove this in the spirit of RFCs being orthogonal. I also think we have too many un-merged RFCs to have half-resolved references between them.

Author: Point #1 - fair.
Point #2 - the other needs approving first!

rfc#0000x described the following Enarx use case states for the default

> Enarx trust model (TODO: check correctness in rfc#0000x and normalise):

Reviewer 2: Should be RFC#00003 if numbers don't change.

 - State 0 - initial state
 - State 1 - pre-attestation
 - State 2 - attested
 - State 3 - provisioned

Non-Enarx use cases were also discussed, but these are irrelevant to this

> discussion, and therefore ignored in this RFC.

Reviewer 3: Same comment as above, let's remove this information cited from another RFC.

Author: It was intended to set the scene.  If we _don't_ have references to other RFCs, how do you know to what to refer?  I don't think you can have it both ways...

The changes between states were characterised as the transition of

> particular components between trust domains:

Reviewer 2: ```suggestion
The changes between states were characterised as the transition of
particular components between different trust domains:
```
Different is implied, but making it explicit may help.

Author: Taken.

 1. host owner trust domain;
 2. tenant trust domain.

This RFC uses the same states, trust domains and components.  Although
the level of description may feel overly exhaustive, the members of the
Enarx project believe that it is important to be explicit about the trust
models being implemented, and how they reflect the "real world", in order
to allow designers, coders, documenters, testers, auditors and users of
the project to have as full an understanding of the process and
architecture as possible.

### How the trust domain change happens
The key state change in terms of the Enarx default trust model is from

> state 1 to state 2.  It is at this point that at **TEE instance** on the

Reviewer 2: ```suggestion
state 1 to state 2.  It is at this point that a **TEE instance** on the
```

host (and in the **host owner's trust domain**), loaded with an **Enarx
runtime image** (which then executes, to become an **Enarx runtime

> instance**) becomes an **Empty Keep**, within the **tenant trust domain**.

Reviewer 2: ```suggestion
instance**) becomes an **Empty Keep**, and moves to the **tenant trust domain**.
```

Might be helpful to make explicit that once it is in the tenant trust domain it is (by definition) no longer in the host owner's trust domain. "moves" might not be the right term, but something that expresses this exclusive nature would be useful, i believe.

Author: Taken, with further changes.

This is achieved by the use of a __trust pivot__: the **host CPU + CPU
firmware**.  The attestation process described in the State 1->2 change
section explicitly notes that that cryptographic identity of the **host
CPU + CPU firmware** must be validated as part of the attestation
measurement check process.

The ability for the tenant to trust the host CPU + CPU firmware underpins
the entire trust model process, and it is vital that all implementations
check it very carefully.

### State 0->1 change

#### State 0 - initial state

>  - host owner trust domain

Reviewer 3: The bullet-pointed lists for all of the state descriptions have inconsistent capitalization. I think they should just be capitalized as complete sentences.

Author: I'll review.

Author: I've lower-cased "Other ..." throughout, I think.  The others name entities, with the exception of host.

    - host

>     - host CPU + CPU firmware

Reviewer 3: Seems a little redundant. When will we have one of these without the other?

Author: We won't (I suspect!), but I'm trying to be as explicit at possible.

Author: And, more important, we're utilising a combination of the capabilities that the two entities provide.

    - Enarx host agent
    - Enarx runtime image
    - Other workloads
 - tenant trust domain
    - Orchestrator
    - Enarx client agent
    - Tenant workload image

>     - Tenant workload image repository

Reviewer 2: ```suggestion
 - Host owner trust domain
    - Host
    - Host CPU + CPU firmware
    - Enarx host agent
    - Enarx runtime image
    - Other workloads
 - Tenant trust domain
    - Orchestrator
    - Enarx client agent
    - Tenant workload image
    - Tenant workload image repository
```
Capitilasing for consistency.

#### State 1

 - host owner trust domain
    - host
    - host CPU + CPU firmware
    - Enarx host agent
    - **TEE instance**
    - **Enarx runtime instance (within TEE instance)**
    - Other workloads
 - tenant trust domain
    - Orchestrator
    - Enarx client agent
    - Tenant workload image

>     - Tenant workload image repository

Reviewer 2: ```suggestion
 - Host owner trust domain
    - Host
    - Host CPU + CPU firmware
    - Enarx host agent
    - **TEE instance**
    - **Enarx runtime instance (within TEE instance)**
    - Other workloads
 - Tenant trust domain
    - Orchestrator
    - Enarx client agent
    - Tenant workload image
    - Tenant workload image repository
```
Capitilasing for consistency.

#### Changes

The changes between state 0 and state 1 are the creation of a TEE instance
and the loading of the Enarx runtime image into that instance.

> The initialisation of the process is prompted by **Orchestrator**, which

Reviewer 2: the Orchestrator, an Orchestrator?

communicates with the **Enarx client agent**, which contacts the **Enarx
host agent** over a network connection (which does not need to be
encrypted).  A mechanism (such as a cryptographic nonce), allowing the
Enarx client agent to match sessions and requests, is included in the
initial request.  The **Enarx host agent** then requests the creation,

> which is performed by the **host CPU + CPU firmware**.

Reviewer 2: Could we explain why the connection between Enarx client agent and Enarx host agent needn't be encrypted? On the face of it, it sounds surprising for a project that is security-focused. Of course, there's a good reason for that, but i think we should make it explicit: the Enarx host agent is merely acting as a proxy, and is not in a position to falsify information from either the CPU + CPU firmware, or the Enarx client agent. Again as is the case with TEEs in general at this point in time, in our CIA model, only availability can be attacked by the host.

Reviewer: This is a good point @axelsimon

The **Enarx runtime image** is typically stored on host storage (local or
remote - the distinction is not relevant to the process).  The loading
process varies between architectures (TODO: is this true?), but is
requested by the **Enarx host agent**.  Although the initial image may
persist on host storage, there is no requirement for or against this,
and so the component is not listed separately in state 1.  We follow
this convention for later states.

Once the **TEE instance** has been created, the **host CPU + CPU firmware**
load the **Enarx runtime image** into the **TEE instance**.  The
**Enarx runtime image**, now loaded, is executed, and becomes the
**Enarx runtime instance**, though it is not yet attested.

As part of the initialisation process, the **Orchestrator** provides the
**Enarx client agent** with one of the following:
 1. access to the **Tenant workload image repository** and a reference to

>  **Tenant workload image** to be provisioned;

Reviewer 2: ```suggestion
 1. access to the **Tenant workload image repository** and a reference to
 the **Tenant workload image** to be provisioned;
```

 2. direct access to the **Tenant workload image**, either as a binary in
 memory or in storage;
 3. a reference allowing the **Enarx client agent** to recontact the

>  **Orchestrator** in the event of successful attestation.

Reviewer 2: I think i don't understand something here, it would seem to me that #3 dependent is on #1 or #2? Or at least, not related: one would want the Orchestrator to be able to hear back from the Enarx client agent if attestation was successful or not in any case.
Or does option 3 mean that the Orchestrator would wait for the Enarx client agent to show the (empty) Keep has been properly attested before providing with an image of the workload to load into the Keep?

Coming back to this after more reading, it appears this is explained at line 295 and following. We should either remove this paragraph here, say it will be further explained below or add something like:
```suggestion
 3. a reference allowing the **Enarx client agent** to recontact the
 **Orchestrator** in the event of successful attestation in order to
 then obtain the **Tenant workload image**.
```

 
Though the encryption of the channel between the Enarx client agent
and the Enarx host agent is not **required**, the following properties are
considered to be **desirable**:

 - authentication of the Enarx client agent to the Enarx host agent
   - this allows the host to control which parties are permitted to create
   Enarx Keeps.  While this has no direct benefit to the tenant, an
   indirect benefit may be that the host can avoid DoS attacks, thereby
   maintaining availability for workload placement and execution.
   - note that the Enarx host agent is *not* trusted by the tenant, and
   it is therefore the responsibility of the host to decide whether to
   trust any such authentication.
 - integrity of the request
   - some DoS attacks or protocol attacks may be feasible, and mitigated
   by integrity protection.
 - encryption of the protocol handshake and further steps
   - some attacks based around protocol versioning or replays may be

>    feasible, and can be avoided by encryption.

Reviewer 2: Often, encryption downgrade attacks are mitigated by enforcing the server-side choice of ciphers (at least, that's what i've often seen with regards to webserders). Would it be similar for us or not?
Our basic approach is that the trust and security are looked at from the perspective of the tenant first and foremost, so we'd want client/tenant-side cipher choice to win over server/host-side. But i'm not sure if we can gain something, even for the tenant, by letting the host decide on ciphers.

For the reasons noted above, encrypting the channel will sometimes be considered
the simplest option.

### State 1->2 change

#### State 1 - pre-attestation

 - host owner trust domain
    - host
    - host CPU + CPU firmware
    - Enarx host agent
    - TEE instance
    - Enarx runtime instance (within TEE instance)
    - Other workloads
 - tenant trust domain
    - Orchestrator
    - Enarx client agent
    - Tenant workload image

>     - Tenant workload image repository

Reviewer 2: ```suggestion
 - Host owner trust domain
    - Host
    - Host CPU + CPU firmware
    - Enarx host agent
    - TEE instance
    - Enarx runtime instance (within TEE instance)
    - Other workloads
 - Tenant trust domain
    - Orchestrator
    - Enarx client agent
    - Tenant workload image
    - Tenant workload image repository
```
Capitilasing for consistency.

#### State 2 - attested

 - host owner trust domain
    - host
    - host CPU + CPU firmware
    - Enarx host agent
    - Other workloads
 - tenant trust domain
    - Orchestrator
    - Enarx client agent
    - **Empty Keep**
    - Tenant workload image

>     - Tenant workload image repository

Reviewer 2: ```suggestion
 - Host owner trust domain
    - Host
    - Host CPU + CPU firmware
    - Enarx host agent
    - Other workloads
 - Tenant trust domain
    - Orchestrator
    - Enarx client agent
    - **Empty Keep**
    - Tenant workload image
    - Tenant workload image repository
```
Capitilasing for consistency.

measurement is correct.  This process includes checks for:

 - host CPU + CPU firmware validity
 - cryptographic correctness
 - TEE validity
 - valid Enarx runtime image.

If any of these checks fail, the state progression is stopped, and the
**Enarx client agent** logs an error and may send an error message to
the **Enarx host agent**.  The **Enarx client agent** also sends an
error message to the Orchestrator.

If the attestation measurement meets the checking criteria, the **Enarx
client agent** retrieves or derives the session key from the previous
communication originating in the **host CPU + CPU firmware** (the exact
mechanism being CPU architecture-dependent).  At this point, the state
transition is considered complete, and the TEE instance, loaded with
the Enarx runtime image, is now an **Empty Keep**, which includes a
running **Enarx runtime instance**. 

### State 2->3 change

#### State 2 - attested

 - host owner trust domain
    - host
    - host CPU + CPU firmware
    - Enarx host agent
    - Other workloads
 - tenant trust domain
    - Orchestrator
    - Enarx client agent
    - **Empty Keep**
    - Tenant workload image

>     - Tenant workload image repository

Reviewer 2: ```suggestion
 - Host owner trust domain
    - Host
    - Host CPU + CPU firmware
    - Enarx host agent
    - Other workloads
 - Tenant trust domain
    - Orchestrator
    - Enarx client agent
    - **Empty Keep**
    - Tenant workload image
    - Tenant workload image repository
```
Capitilasing for consistency.

#### State 3 - provisioned

 - host owner trust domain
    - host
    - host CPU + CPU firmware
    - Enarx host agent
    - Other workloads
 - tenant trust domain
    - Orchestrator
    - Enarx client agent
    - **Provisioned Keep**
    - Tenant workload image

>     - Tenant workload image repository

Reviewer 2: ```suggestion
 - Host owner trust domain
    - Host
    - Host CPU + CPU firmware
    - Enarx host agent
    - Other workloads
 - Tenant trust domain
    - Orchestrator
    - Enarx client agent
    - **Provisioned Keep**
    - Tenant workload image
    - Tenant workload image repository
```
Capitilasing for consistency.

components, protocols or other artifacts.

When a trade-off between the security of the tenant and workload needs
to be made against the security of any other party (including host owner),
the security of the tenant MUST always win.

When measures for a party other than the tenant are being considered,
the general principle should be to choose the highest levels of security,
but any choices MUST NOT compromise tenant security. 

The Tenant Workload MUST never be unencrypted on the network (in transit)
within the Enarx-managed process.

The Tenant Workload MUST never be unencrypted in storage once it has been
transmitted by the Enarx client agent.

The Tenant Workload MUST never be unencrypted in RAM once it has been
transmitted by the Enarx client agent.

The Tenant Workload MUST never be available in unencrypted form on the host.

### Other requirements
The channel between the Enarx client agent and the Enarx host agent MAY be
encrypted.

The Enarx client agent SHOULD authenticate to the Enarx host agent.

Integrity protection of the request SHOULD be applied.

The protocol handshake and further steps SHOULD be encrypted.

> If integrity protection is to the protocol handshake and further
> communications between the Enarx client agent and the Enarx host agent
> are applied, all detected integrity failures SHOULD be reported to
> the Enarx client agent via the Enarx host agent where possible.

Reviewer 2: Unless this is a grammatical turn of phrase i'm unfamiliar with, it seems simpler to me to say:
```suggestion
If integrity protection is applied to the protocol handshake and to further
communications between the Enarx client agent and the Enarx host agent,
all detected integrity failures SHOULD be reported to the Enarx client agent 
via the Enarx host agent where possible.
```

Author: Fixed.

The integrity of the tenant's trust domain MUST always be given the
highest priority.

Protections for the tenant MUST always be prioritised above those of
other parties such as the host.

Confidentiality of the tenant's workload MUST be protected.

Integrity of the tenant's workload MUST be protected.

All protocol errors SHOULD be logged by the Enarx client agent.

All attestation measurements MUST be cryptographically signed.

All attestation measurements MUST be checked.

All attestation measurements SHOULD be logged by the Enarx client agent.

All Enarx client agent logs SHOULD be integrity protected.

All attestation measurement failures MUST result in a cessation of
the standard state progression.

The cryptographic validity of the host CPU + CPU firmware MUST be
checked in every attestation measurement check.

> All cryptographic validity check failures of the host CPU + CPU firmware

Reviewer 2: ```suggestion
Any and all cryptographic validity check failures of the host CPU + CPU firmware
```

agent.

All attestation measurement failures SHOULD result in an error message
being sent to the Orchestrator.

The Enarx client agent MUST generate and include session tracking
information in the initial communication with the Enarx host agent.

The session tracking information MUST be included in all communications
between the Enarx client agent and the Enarx host agent until the Keep
is fully provisioned.

The Tenant Workload MUST never be unencrypted on the network (in transit)
within the Enarx-managed process.

The Tenant Workload MUST never be unencrypted in storage once it has been
transmitted by the Enarx client agent.

The Tenant Workload MUST never be unencrypted in RAM once it has been
transmitted by the Enarx client agent.

The Tenant Workload MUST never be available in unencrypted form on the host.

The Enarx client agent MUST be the component which performs attestation
measurement checks.

The Enarx client agent MAY pass attestation measurements to the Orchestrator.

The Enarx client agent MUST NOT accept override instructions from the
Orchestrator to proceed with deployment of a Tenant workload image in the

> case of an attestation measurement failure.

Reviewer 2: This is worth saying, as it reminds of the central role (in terms of trust domains) that the Enarx client agent holds.
